### PR TITLE
Refactoring the compact_partition to support different object store implementations

### PR DESCRIPTION
**Pull request by raghumdani**
#148 

- Tested all implementations on internal 15TB table
- Unit test pending for each object store impl. 

update: unit tests added

### CODE REVIEW

File: deltacat/io/redis_object_store.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+            cache_result = client.mget(uids)
+            assert len(cache_result) == len(
+                uids
+            ), "Not all values were returned from cache"
+
+            total_bytes = 0
+
+            deserialize_start = time.monotonic()
+            for serialized in cache_result:
+                deserialized = cloudpickle.loads(serialized)
```
Comment (valiantljk) [inline deltacat/io/redis_object_store.py:77]: This assumes that the object is cloudpickled first, which may not be true

**Review by valiantljk — COMMENTED** [review]
Great work!
Major concern: other than plasma object store, how does other 'object store' solutions handle object recovery in case of failure? And ray has the locality-aware scheduling, so that tasks can be placed close to object, I assume this object ref resolution only works on plasma object store; Thirdly, when the task and object are not physically close, ray will trigger the inter-node object transfer, how does the new proposed object store path handle this?

File: deltacat/compute/compactor/steps/dedupe.py [hunk 2]
```diff
@@ -114,15 +116,12 @@ def _timed_dedupe(
...
-            cloudpickle.loads(obj_id_pkl) for obj_id_pkl in object_ids
-        ]
         logger.info(
             f"[Dedupe task {dedupe_task_index}] Getting delta file envelope "
-            f"groups for {len(src_file_records_obj_refs)} object refs..."
+            f"groups for {len(object_ids)} object refs..."
         )
 
-        delta_file_envelope_groups_list = ray.get(src_file_records_obj_refs)
+        delta_file_envelope_groups_list = object_store.get_many(object_ids)
```
[the remark this answers is not in the document]
Comment (raghumdani) [reply deltacat/compute/compactor/steps/dedupe.py:124]: no. There is a default set in `compaction_session.py` and it doesn't change current behavior.

File: deltacat/io/object_store.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        of object references corresponding to each object in the input.
+        """
+
+    def get(self, ref: Any, *args, **kwargs) -> object:
+        """
+        Get a single object from an object store.
+        """
+        return self.get_many([ref])[0]
+
+    def get_many(self, refs: List[Any], *args, **kwargs) -> List[object]:
```
[the remark this answers is not in the document]
Comment (raghumdani) [reply deltacat/io/object_store.py:37]: it's mainly to bring all object stores under a common interface. Implementation wise, we will use `ray.get` to retrieve all objects.

File: deltacat/io/redis_object_store.py (hunk 1 above)
Comment (raghumdani) [reply deltacat/io/redis_object_store.py:77]: you can only get what you put. So, from the `put` and `put_many` implementation of `RedisObjectStore`, it should be always true.

**Comment by raghumdani** [comment]
> other than plasma object store, how does other 'object store' solutions handle object recovery in case of failure?

Can you explain the failure scenario? Behavior wise, it should be same. 

> And ray has the locality-aware scheduling, so that tasks can be placed close to object, I assume this object ref resolution only works on plasma object store

Memcached and Redis both put the object into the node they were called in. Did you mean any other feature we won't make use of?

> when the task and object are not physically close, ray will trigger the inter-node object transfer, how does the new proposed object store path handle this?

each object store is handling it differently, for example 1) S3 is global, 2) same file system must be attached to all instances, 3) memcached and redis will bring in object from it's owner node, 4) ray handles owner resolution for us.

File: deltacat/io/redis_object_store.py (hunk 1 above)
Comment (valiantljk) [reply deltacat/io/redis_object_store.py:77]: what if later we use this object store interface to put un-pickled object?

**Comment by valiantljk** [comment]
> > other than plasma object store, how does other 'object store' solutions handle object recovery in case of failure?
> 
> Can you explain the failure scenario? Behavior wise, it should be same.
> 
features like 'object reconstruction' seems to be missing in the proposed new interface. Failures can be any kind, e.g., node failure, owner failure, etc: https://docs.ray.io/en/latest/ray-core/fault_tolerance/objects.html
> > And ray has the locality-aware scheduling, so that tasks can be placed close to object, I assume this object ref resolution only works on plasma object store
> 
> Memcached and Redis both put the object into the node they were called in. Did you mean any other feature we won't make use of?
Yeah, when we pass the ref to tasks, it'll be auto-resolved, we don't need to call 'get' explicitly, and object transfer will be triggered underneath. Also, what if memory is full and disk spilling is needed?
> 
> > when the task and object are not physically close, ray will trigger the inter-node object transfer, how does the new proposed object store path handle this?
> 
> each object store is handling it differently, for example 1) S3 is global, 2) same file system must be attached to all instances, 3) memcached and redis will bring in object from it's owner node, 4) ray handles owner resolution for us.

I realize this question is same with the object ref resolution, so if we have this new object store interface, we'll assume future dev will always explicitly call get/get_many to retrieve the objects. This seems to be against some of the Ray object store design?

File: deltacat/io/redis_object_store.py (hunk 1 above)
Comment (raghumdani) [reply deltacat/io/redis_object_store.py:77]: That will be part of the implementation whether to pickle or not. The `object_store.get` will return what was `object_store.put`

**Comment by raghumdani** [comment]
> features like 'object reconstruction' seems to be missing in the proposed new interface. Failures can be any kind, e.g., node failure, owner failure, etc: https://docs.ray.io/en/latest/ray-core/fault_tolerance/objects.html

object reconstruction is a ray feature that works on task returns in the ray plasma store when a worker dies. However, if we call ray.put explicitly we are responsible for reconstructing the object anyway. What is the behavior or feature are you looking for?

> Yeah, when we pass the ref to tasks, it'll be auto-resolved, we don't need to call 'get' explicitly, and object transfer will be triggered underneath. Also, what if memory is full and disk spilling is needed?

This is a ray feature, not an object store feature. Moreover, for our use-case we are circumventing this by pickling the object ref which is an anti-pattern.

File: deltacat/compute/compactor/steps/materialize.py [hunk 4]
```diff
@@ -96,10 +78,13 @@ def _stage_delta_from_manifest_entry_reference_list(
             delta_type == DeltaType.UPSERT
         ), "Stage delta with existing manifest entries only supports UPSERT delta type!"
         manifest = Manifest.of(entries=manifest_entry_list_reference, uuid=str(uuid4()))
-        delta = _stage_delta_implementation(
-            data=manifest,
-            partition=partition,
-            stage_delta_from_existing_manifest=True,
+        delta = Delta.of(
```
[the remark this answers is not in the document]
Comment (raghumdani) [reply deltacat/compute/compactor/steps/materialize.py:81]: Yes, that was intentional as it's removes the redundant method. We can anyway modify that when we implement `stage_delta`.

**Comment by valiantljk** [comment]
LGTM. Please note down any side effect, e.g., any difference between the new object store and the default plasma store.

**Review by valiantljk — APPROVED** [review]

**Comment by Zyiqin-Miranda** [comment]
LGTM

**Pull request merged**